### PR TITLE
bpo-38691 Added a switch to ignore PYTHONCASEOK when -E or -I flags passed

### PR DESCRIPTION
**Pull request by idomic**
https://bugs.python.org/issue38691

### CODE REVIEW

**Comment by idomic** [comment]
@vstinner Added a PR 👍

**Comment by idomic** [comment]
Not sure why the test_importlib test fails, I think it expects the value no matter what?

**Review by corona10 — CHANGES_REQUESTED** [review]
@idomic 

Please run `make regen-all` before commit.
You can see some diffs which are generated from this command.
And please commit it after do this.

Thanks for the contribution

File: Lib/importlib/_bootstrap_external.py [hunk 1]
```diff
@@ -34,8 +34,13 @@ def _make_relax_case():
             key = b'PYTHONCASEOK'
 
         def _relax_case():
-            """True if filenames must be checked case-insensitively."""
-            return key in _os.environ
+            """True if filenames must be checked case-insensitively.
+            If the -E or -I flags are used PYTHONCASEOK is ignored
+            """
+            if sys.flags.ignore_environment:
```
Comment (corona10) [inline Lib/importlib/_bootstrap_external.py]:
```python
return not sys.flags.ignore_environment and key in _os.environ
```

File: Misc/NEWS.d/next/Library/2020-02-02-14-48-01.bpo-38691.ewUjZn.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+updated importlib._bootstrap_external._make_relax_case to ignore PYTHONCASEOK when -E or -I flags are used.
```
Comment (corona10) [inline Misc/NEWS.d/next/Library/2020-02-02-14-48-01.bpo-38691.ewUjZn.rst]:
```suggestion
Updated :meth:`importlib._bootstrap_external._make_relax_case` to ignore :envvar:`PYTHONCASEOK` when -E or -I flags are used.
```

**Review by corona10 — CHANGES_REQUESTED** [review]

**Comment by vstinner** [comment]
>  corona10 requested a review from vstinner 2 days ago 

I will not review this change: most CIs are red. Please fix it before I can look at this PR.

File: Lib/importlib/_bootstrap_external.py (hunk 1 above)
Comment (idomic) [reply Lib/importlib/_bootstrap_external.py]: Accepted both comments thank you!

**Comment by idomic** [comment]
> @idomic
> 
> Please run `make regen-all` before commit.
> You can see some diffs which are generated from this command.
> And please commit it after do this.
> 
> Thanks for the contribution

So I ran those 3, now I have a conflict on the importlib_external.h file:

git clean -fdx
./configure --with-pydebug
make regen-all

**Comment by corona10** [comment]
@idomic 
Please rebase before make regen-all

**Comment by idomic** [comment]
@corona10  I think I might be doing something wrong.

I was cleaning the folder using: git clean -fdx
Then rebasing master into current, running ./configure --with-pydebug and make regen-all.
Then I push to remote branch and letting it merge with remote.
I have also tried to merge just this conflicted importlib file from master and take the master's version but it failed the CI.

**Comment by idomic** [comment]
Adding @taleinat to accelerate PR

**Comment by vstinner** [comment]
This PR looks overcomplicated. It has 24 commits but only changes 1 line of code... Maybe create a new PR on top of the up to date master branch, and ensure that you run "make regen-all" to update  Python/importlib_external.h.

**Comment by taleinat** [comment]
Hi @idomic!

> I think I might be doing something wrong.

Yep, it seems that you've become a bit tangled up with the git merges and rebases. The final step that caused the current mess was when, after rebasing, you allowed merging with the remote branch. Instead, you should have force-pushed the rebased branch, using `git push -f` (use with care!).

I suggest you (hard-) reset your branch to the commit before merging the remote branch, and then force-push it.

(Generally, for branches with an open PR, I always recommend merging rather than rebasing. This is for several reasons, one of them being avoiding issues like this. Even when you're asked "please rebase", merging will usually achieve the same end result in terms of the files, but make following and managing the PR simpler.)

**Comment by idomic** [comment]
Thanks for the quick help @taleinat 

So Once hard resetting and cherry picking the CI starts running, windows x64 fails on:
"D:\a\cpython\cpython\PCbuild\_freeze_importlib.vcxproj(146,5): error : If you are not developing on Windows but you see this error on a continuous integration build, you need to run 'make regen-all' and commit any changes."

Once I run configure with --pydebug and make regen-all it creates a conflict in the file importlib_external.h even though I'm forcing my local branch's push.

File: Misc/NEWS.d/next/Library/2020-02-11-13-01-38.bpo-38691.oND8Sk.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Updated :meth:`importlib._bootstrap_external._make_relax_case` to ignore :envvar:`PYTHONCASEOK` when -E or -I flags are used.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2020-02-11-13-01-38.bpo-38691.oND8Sk.rst]: Please avoid documenting private modules/functions in public Changelog entries. I propose:

```
The :mod:`importlib` module now ignore the :envvar:`PYTHONCASEOK`
environment variable when :option:`-E` or :option:`-I` command line option is used.
```

Please document also this incompatible change at:

* https://docs.python.org/dev/library/functions.html#__import__
* https://docs.python.org/dev/whatsnew/3.9.html#changes-in-the-python-api

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by idomic** [comment]
Thanks for the comments @vstinner 
I have made the requested changes; please review again

File: Doc/library/functions.rst [hunk 4]
```diff
@@ -1826,6 +1826,10 @@ are always available.  They are listed here in alphabetical order.
       Negative values for *level* are no longer supported (which also changes
       the default value to 0).
 
+   .. versionchanged:: 3.9
+      When the command line options -E or -I is used, the environment variable
+      PYTHONCASEOK is ignored.
```
Comment (vstinner) [inline Doc/library/functions.rst]: I suggest to add "now", suggestion which also added `:envvar:`  markup:

```
When the command line options :option:`-E` or :option:`-I` is used,
the environment variable :envvar:`PYTHONCASEOK` is now ignored.
```

File: Doc/whatsnew/3.9.rst [hunk 5]
```diff
@@ -262,6 +262,8 @@ Changes in the Python API
 * The :mod:`venv` activation scripts no longer special-case when
   ``__VENV_PROMPT__`` is set to ``""``.
 
+* The :mod:`importlib` module now ignore the :envvar:`PYTHONCASEOK`
```
Comment (vstinner) [inline Doc/whatsnew/3.9.rst]:
```suggestion
* The :mod:`importlib` module now ignores the :envvar:`PYTHONCASEOK`
```

File: Misc/NEWS.d/next/Library/2020-02-11-13-01-38.bpo-38691.oND8Sk.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+The :mod:`importlib` module now ignore the :envvar:`PYTHONCASEOK`
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2020-02-11-13-01-38.bpo-38691.oND8Sk.rst]:
```suggestion
The :mod:`importlib` module now ignores the :envvar:`PYTHONCASEOK`
```

File: Doc/whatsnew/3.9.rst [hunk 7]
```diff
@@ -262,6 +262,8 @@ Changes in the Python API
...
+  environment variable when :option:`-E` or :option:`-I` command line option is used.
```
Comment (taleinat) [inline Doc/whatsnew/3.9.rst]: Suggestion for better English:

```suggestion
  environment variable when the :option:`-E` or :option:`-I` command line options
  are used.
```

**Comment by idomic** [comment]
I have made the requested changes; please review again

File: Doc/whatsnew/3.9.rst (hunk 5 above)
Comment (vstinner) [reply Doc/whatsnew/3.9.rst]: > I have made the requested changes; please review again

Well, you didn't update this one. English is not my first language, so I'm not sure, but it should be "The module ignoreS" here, no? with an S.

Comment (taleinat) [reply Doc/whatsnew/3.9.rst]: Yep, you're right regarding the English grammar here, @vstinner.

File: Doc/library/functions.rst [hunk 8]
```diff
@@ -1826,6 +1826,10 @@ are always available.  They are listed here in alphabetical order.
       Negative values for *level* are no longer supported (which also changes
       the default value to 0).
 
+   .. versionchanged:: 3.9
+      When the command line options :option:`-E` or :option:`-I` is used,
```
Comment (taleinat) [inline Doc/library/functions.rst]:
```suggestion
      When the command line options :option:`-E` or :option:`-I` are used,
```

**Comment by taleinat** [comment]
Also, note that when you're done, you'll have to merge and regenerate `Python/importlib_external.h` one last time.

File: Doc/library/functions.rst (hunk 8 above)
Comment (idomic) [reply Doc/library/functions.rst]: Will fix both comments now

**Comment by idomic** [comment]
I have made the requested changes; please review again

File: Doc/library/functions.rst (hunk 8 above)
Comment (idomic) [reply Doc/library/functions.rst]: I understood what the problem was, apparently I misconfigured my local repo so there was no up stream, now everything should pass fine.

**Comment by vstinner** [comment]
@idomic: Thanks, I merged your PR.

**Comment by pablogsal** [comment]
@vstinner This PR is likely to be the cause of https://buildbot.python.org/all/#/builders/275/builds/245

**Comment by vstinner** [comment]
> @vstinner This PR is likely to be the cause of https://buildbot.python.org/all/#/builders/275/builds/245

Right: https://bugs.python.org/issue38691#msg362202

**Pull request merged**